HPCC-29923 Avoid CRowServer cancel if not running

jakesmith opened the pull request:
This avoids an inherent delay being introduced by the MP layer.

CRowServer is used by Global Group, Global TopN and a refactoring
form of IndexWrite (where is merges down to less parts).

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith committed:
HPCC-29923 Avoid CRowServer cancel if not running

This avoids an inherent delay being introduced by the MP layer.

CRowServer is used by Global Group, Global TopN and a refactoring
form of IndexWrite (where is merges down to less parts).

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

ghalliday commented on thorlcr/thorutil/thormisc.cpp:
```diff
@@ -1371,8 +1371,12 @@ class CRowServer : public CSimpleInterface, implements IThreaded, implements IRo
+        bool wanted = true;
+        if (running.compare_exchange_strong(wanted, false))
+        {
+            running = false;
```
trivial: not needed (it will be set to false by the compare_exchange

jakesmith replied:
removed.

ghalliday reviewed (commented):
@jakesmith One comment - looks good.
Is 9.2.x the correct target?

jakesmith commented:
> Is 9.2.x the correct target?

I figured that it's probably safe, but a very long standing issue that doesn't need to go into 9.0.x, so 9.2.x was more appropriate,
but happy to change it to 9.0.x if you think it would be good to?

ghalliday commented:
@jakesmith does it only affect groups?  It is a bug fix which is relatively trivial so I think 9.0.x would be appropriate.

@GordonSmith if Jake agrees with my comment and removes the assignment then it can be merged.

jakesmith commented:
> @jakesmith does it only affect groups? It is a bug fix which is relatively trivial so I think 9.0.x would be appropriate.

I've updated the commit/PR message, but as per the JIRA title, it is also used in Global TopN and a unusual indexwrite case.

> 
> @GordonSmith if Jake agrees with my comment and removes the assignment then it can be merged.

unnecessary assignment removed.
Rebased to 9.0.x

jakesmith commented:
> @GordonSmith if Jake agrees with my comment and removes the assignment then it can be merged.

@GordonSmith - rebased to 9.0.x - I think you can merge it.
